test/test_ml.ml

```diff
@@ -534,24 +534,26 @@ let testCreateRef () =
   React.Ref.setCurrent reactRef (Js_of_ocaml.Js.Opt.return 1) ;
   assert_equal (React.Ref.current reactRef) (Js_of_ocaml.Js.Opt.return 1)
 
-(* TODO: figure out the `forwardRef` API *)
-let testForwardRef () = ()
-(* let testForwardRef () = *)
-(*   let module FancyButton = struct *)
-(*     let make = *)
-(*       React.Dom.forwardRef (fun ~children ref -> *)
-(*           (button [|ref_ ref; className "FancyButton"|] children) ) *)
-(*   end in *)
-(*   withContainer (fun c -> *)
-(*       let count = ref 0 in *)
-(*       let buttonRef = *)
-(*         React.Dom.Ref.callbackDomRef (fun _ref -> count := !count + 1) *)
-(*       in *)
-(*       act (fun () -> *)
-(*           React.Dom.render *)
-(*             (FancyButton.make ~ref:buttonRef [div [||] []]) *)
-(*             (Html.element c) ) ; *)
-(*       assert_equal !count 1 ) *)
+let testForwardRef () =
+  let module FancyButton = struct
+    let make ~ref children =
+      (* TODO: This doesn't seem to test forwardRef, whatever it does... *)
+      (* React.Dom.forwardRef (fun children ref -> *)
+      div [||]
+        [ button [|ref_ ref; className "FancyButton"|] children
+        ; button [|ref_ ref; className "FancyButton"|] children ]
+    (* ) *)
+  end in
+  withContainer (fun c ->
+      let count = ref 0 in
+      let buttonRef =
+        React.Dom.Ref.callbackDomRef (fun _ref -> count := !count + 1)
+      in
+      act (fun () ->
+          React.Dom.render
+            (FancyButton.make ~ref:buttonRef [div [||] []])
+            (Html.element c) ) ;
+      assert_equal !count 2 )
```

Comment:
I still don't understand the `forwardRef` API, or why it's needed at all. Seems to work just fine without it...

Reply:
I think I can explain.

`forwardRef` main goal is to "pass over" the ref passed to a call to `createElement` (where React takes over) to the component implementation itself. It's well explained here: https://reactjs.org/docs/forwarding-refs.html.

When we migrated to OCaml syntax in #72, the `React.createElement` call was moved from the consumers of the component, to the component implementation itself, so that one could create elements by just calling `Foo.make`. 

This means that now, when a consumer passes a `ref` to some uppercase component, this component will receive exactly that same `ref`,  without React interceding at any point when this happens (it just a call to `make`).

All this means that there is no need anymore for something like `forwardRef` anymore. I will open an issue to remove it as it only adds confusion.

Let me know if this makes sense :) i'm not sure the explanation is really great 😆

Reply:
Ah, yeah I think that sort of makes sense. Thanks!

I'll rewrite the OCaml test to just check that `~ref` is forwarded then. But beware that the Reason test still doesn't check that the ref is actually forwarded, whether `forwardRef` is used or not.

> When we migrated to OCaml syntax in #72, the `React.createElement` call was moved from the consumers of the component, to the component implementation itself, so that one could create elements by just calling Foo.make.

I wonder what's lost by doing this. Maybe something with higher-order components? Or some dynamic behaviour we don't care about? Surely there's a reason it was done this way.

Reply:
> I'll rewrite the OCaml test to just check that ~ref is forwarded then. But beware that the Reason test still doesn't check that the ref is actually forwarded, whether forwardRef is used or not.

I think the bug is that we're not passing `ref` to the `createElement` call :(

https://github.com/ml-in-barcelona/jsoo-react/blob/5fbcb1dbe8d5b8b7bbf75ee9909106e17fe508f0/ppx/test/pp_ocaml.expected#L37-L38

We should pass optional `ref` over, like we do with `key`.

I will revisit this when I have some time.

> I wonder what's lost by doing this. Maybe something with higher-order components? Or some dynamic behaviour we don't care about? Surely there's a reason it was done this way.

The only downside that I could think of is that we lose the ability to leverage React warnings for `key` prop missing for children:

```
Warning: Each child in a list should have a unique "key" prop.
```

These warnings are shown in development mode when children are built from expressions that are not literal arrays. But in any case, these checks are not very accurate and can easily lead to false positives, e.g.

```js
export default function App() {
  const children = [
    <h1>Foo</h1>,
    <h2>Bar</h2>
  ];
  return <div>{children}</div>;
}
```

This will trigger the warning, but it's obvious in this case that `key` is not needed because the array has constant size.

The check itself is quite primitive, it only checks if the element is a react element by [checking the existence of a `_store` property](https://github.com/facebook/react/blob/c1220ebdde506de91c8b9693b5cb67ac710c8c89/packages/react-reconciler/src/ReactChildFiber.new.js#L70).

Besides this, I could not think of anything else.